
version: '3'
services:
  kafka:
    image: 'bitnami/kafka:3.7'
    hostname: kafka
    ports:
      - '9092:9092'
      - '9094:9094'
    environment:
      - KAFKA_CFG_NODE_ID=0
      - KAFKA_CFG_PROCESS_ROLES=controller,broker
      - KAFKA_CFG_CONTROLLER_QUORUM_VOTERS=0@kafka:9093
      - KAFKA_CFG_LISTENERS=PLAINTEXT://:9092,CONTROLLER://:9093,EXTERNAL://:9094
      - KAFKA_CFG_ADVERTISED_LISTENERS=PLAINTEXT://kafka:9092,EXTERNAL://localhost:9094
      - KAFKA_CFG_LISTENER_SECURITY_PROTOCOL_MAP=CONTROLLER:PLAINTEXT,PLAINTEXT:PLAINTEXT,EXTERNAL:PLAINTEXT
      - KAFKA_CFG_CONTROLLER_LISTENER_NAMES=CONTROLLER
      - KAFKA_CFG_INTER_BROKER_LISTENER_NAME=PLAINTEXT
      - KAFKA_AUTO_CREATE_TOPICS_ENABLE=true
    healthcheck:
      test: ['CMD', 'kafka-topics.sh', '--bootstrap-server', 'kafka:9092', '--list']
      interval: 20s
      timeout: 10s
      retries: 5
    volumes:
      - kafka-data:/bitnami/kafka

  data-generator:
    build:
      context: ./data-generator
    depends_on:
      kafka:
        condition: service_healthy
    command: ['python', 'main.py']

volumes:
  kafka-data:
    driver: local
